src/swaps/oneinch/OneinchSwapProvider.js

```diff
@@ -0,0 +1,263 @@
+import { getTxFee } from '../../utils/fees'
+
+const nativeAssetAddress = '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee'
+const chainToRpcProviders = {
+  1: `https://mainnet.infura.io/v3/${buildConfig.infuraApiKey}`,
+  56: 'https://bsc-dataseed.binance.org',
+  137: 'https://rpc-mainnet.matic.network'
+}
+
+class OneinchSwapProvider extends SwapProvider {
+  constructor ({ providerId, agent, routerAddress }) {
+    super(providerId)
+    this.agent = agent
+    this.routerAddress = routerAddress
+  }
+
+  async getSupportedPairs () {
+    return []
+  }
+
+  async getQuote ({ network, from, to, amount }) {
+    if (!isEthereumChain(from) || !isEthereumChain(to) || amount <= 0) return null
+    const fromAmountInUnit = BN(currencyToUnit(cryptoassets[from], BN(amount)))
+    const chainIdFrom = ChainNetworks[cryptoassets[from].chain][network].chainId
+    const chainIdTo = ChainNetworks[cryptoassets[to].chain][network].chainId
+    if (chainIdFrom !== chainIdTo || !chainToRpcProviders[chainIdFrom]) return null
+
+    const trade = await axios({
+      url: this.agent + `/${chainIdFrom}/quote`,
+      method: 'get',
+      params: { fromTokenAddress: assets[from].contractAddress || nativeAssetAddress, toTokenAddress: assets[to].contractAddress || nativeAssetAddress, amount: fromAmountInUnit.toNumber() }
+    })
+    const toAmountInUnit = BN(trade.data.toTokenAmount)
+    return {
+      from,
+      to,
+      // TODO: Amounts should be in BigNumber to prevent loss of precision
+      fromAmount: fromAmountInUnit.toNumber(),
+      toAmount: toAmountInUnit.toNumber()
+    }
+  }
+
+  async approveTokens ({ network, walletId, quote }) {
+    const fromChain = cryptoassets[quote.from].chain
+    const toChain = cryptoassets[quote.to].chain
+    const chainId = ChainNetworks[fromChain][network].chainId
+    if (fromChain !== toChain || !chainToRpcProviders[chainId]) return null
+
+    const api = new ethers.providers.StaticJsonRpcProvider(chainToRpcProviders[chainId])
+    const erc20 = new ethers.Contract(cryptoassets[quote.from].contractAddress, ERC20.abi, api)
+    const fromAddressRaw = await this.getSwapAddress(network, walletId, quote.from, quote.toAccountId)
+    const fromAddress = chains[fromChain].formatAddress(fromAddressRaw)
+    const allowance = await erc20.allowance(fromAddress, this.routerAddress)
+    const inputAmount = ethers.BigNumber.from(BN(quote.fromAmount).toFixed())
+    if (allowance.gte(inputAmount)) {
+      return {
+        status: 'APPROVE_CONFIRMED'
+      }
+    }
+
+    const callData = await axios({
+      url: this.agent + `/${chainId}/approve/calldata`,
+      method: 'get',
+      params: { tokenAddress: cryptoassets[quote.from].contractAddress }
+    })
+
+    const account = this.getAccount(quote.fromAccountId)
+    const client = this.getClient(network, walletId, quote.from, account?.type)
+    const approveTx = await client.chain.sendTransaction(callData.data)
```

Comment:
Some of the parameters from `callData.data` cannot go into the signature for `sendTransaction`
I think it's better to set it explicitly.

Also need to add the fee properly here. in `callData` it's `gasPrice` but the correct parameter is `fee` and it needs to be in GWEI

Reply:
The fee (gas Price in gwei) selected by the user is under `quote.fee`